File: temporal/workflow.go
package temporal

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

// Activity stubs — implementations live in worker/src/activities.ts and worker/src/run-agent.ts.
// Names must match the keys registered in the TypeScript worker exactly.
var (
	yakClaim         = "YakClaim"
	yakRelease       = "YakRelease"
	yakMarkDone      = "YakMarkDone"
	writePRToYak     = "WritePRToYak"
	initWorkspace    = "InitWorkspace"
	cleanupWorkspace = "CleanupWorkspace"
	runAgent         = "RunAgent"
	createDraftPR    = "CreateDraftPR"
	watchPRMerged    = "WatchPRMerged"
)

// YakWorkflow is a single long-running workflow for one yak.
func YakWorkflow(ctx workflow.Context, yakName string, repoRoot string, cfg PiConfig) (string, error) {
	state := &YakWorkflowState{
		YakName: yakName,
		Phase:   "init",
	}

	if err := workflow.SetQueryHandler(ctx, "yak_status", func() (YakWorkflowState, error) {
		return *state, nil
	}); err != nil {
		return "", err
	}

	wontDoChan := workflow.GetSignalChannel(ctx, "wont-do")
	wontDo := false
	workflow.Go(ctx, func(ctx workflow.Context) {
		wontDoChan.Receive(ctx, nil)
		wontDo = true
	})

	noRetry := func(timeout time.Duration) workflow.Context {
		return workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
			StartToCloseTimeout: timeout,
			RetryPolicy:         &temporal.RetryPolicy{MaximumAttempts: 1},
		})
	}
	withRetry := func(timeout time.Duration, attempts int32) workflow.Context {
		return workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
			StartToCloseTimeout: timeout,
			RetryPolicy:         &temporal.RetryPolicy{MaximumAttempts: attempts},
		})
	}

	// 1. Claim the yak.
	state.Phase = "claiming"
	if err := workflow.ExecuteActivity(withRetry(30*time.Second, 3), yakClaim, yakName).Get(ctx, nil); err != nil {
		return "", fmt.Errorf("claim yak %s: %w", yakName, err)
	}

	defer func() {
		if state.Phase != "done" && !wontDo {
			workflow.ExecuteActivity(withRetry(30*time.Second, 3), yakRelease, yakName,
				fmt.Sprintf("workflow interrupted at phase %s", state.Phase)).Get(ctx, nil)
		}
	}()

	// 2. Init workspace.
	state.Phase = "init-workspace"
	var workspaceName string
	if err := workflow.ExecuteActivity(noRetry(60*time.Second), initWorkspace, repoRoot, yakName).Get(ctx, &workspaceName); err != nil {
		return "", fmt.Errorf("init workspace: %w", err)
	}
	state.Workspace = workspaceName

	defer workflow.ExecuteActivity(noRetry(30*time.Second), cleanupWorkspace, repoRoot, workspaceName).Get(ctx, nil)

	// 3. Run Pi agent.
	state.Phase = "implementing"
	if err := workflow.ExecuteActivity(noRetry(4*time.Hour), runAgent, yakName, repoRoot, workspaceName, cfg).Get(ctx, nil); err != nil {
		return "", fmt.Errorf("run agent: %w", err)
	}

	if wontDo {
		return fmt.Sprintf("yak %s marked won't-do during implementation", yakName), nil
	}

	// 4. Create draft PR.
	state.Phase = "creating-pr"
	var pr PRResult
	if err := workflow.ExecuteActivity(noRetry(5*time.Minute), createDraftPR, repoRoot, workspaceName, yakName).Get(ctx, &pr); err != nil {
		return "", fmt.Errorf("create draft PR: %w", err)
	}
	state.PRURL = pr.PRURL
	state.PRNumber = pr.PRNumber

	workflow.ExecuteActivity(withRetry(30*time.Second, 3), writePRToYak, yakName, pr.PRURL).Get(ctx, nil)

	// 5. Wait for merge.
	state.Phase = "waiting-for-merge"
	sel := workflow.NewSelector(ctx)
	mergedCh := workflow.NewChannel(ctx)

	workflow.Go(ctx, func(gCtx workflow.Context) {
		pollCtx := workflow.WithActivityOptions(gCtx, workflow.ActivityOptions{
			StartToCloseTimeout: 7 * 24 * time.Hour,
			HeartbeatTimeout:    2 * time.Minute,
			RetryPolicy:         &temporal.RetryPolicy{MaximumAttempts: 1},
		})
		var merged bool
		if err := workflow.ExecuteActivity(pollCtx, watchPRMerged, pr.PRNumber, repoRoot).Get(gCtx, &merged); err == nil && merged {
			mergedCh.Send(gCtx, true)
		}
	})

	sel.AddReceive(mergedCh, func(c workflow.ReceiveChannel, _ bool) { c.Receive(ctx, nil) })
	sel.AddReceive(wontDoChan, func(c workflow.ReceiveChannel, _ bool) {
		c.Receive(ctx, nil)
		wontDo = true
	})
	sel.Select(ctx)

	if wontDo {
		workflow.ExecuteActivity(withRetry(30*time.Second, 3), yakRelease, yakName, "marked won't-do").Get(ctx, nil)
		return fmt.Sprintf("yak %s marked won't-do", yakName), nil
	}

	// 6. Close the yak.
	state.Phase = "done"
	if err := workflow.ExecuteActivity(withRetry(30*time.Second, 3), yakMarkDone, yakName).Get(ctx, nil); err != nil {
		return "", fmt.Errorf("mark yak done: %w", err)
	}

	return fmt.Sprintf("yak %s done (PR %s)", yakName, pr.PRURL), nil
}

File: worker/src/workflow.ts
import {
  condition,
  defineQuery,
  defineSignal,
  proxyActivities,
  setHandler,
  upsertSearchAttributes,
} from "@temporalio/workflow";
import type * as allActivities from "./activities.js";
import type { PiConfig, PRResult, PRWatchResult, WorkflowConfig, YakWorkflowState } from "./types.js";

// Default max agent run time: 2 hours. The Temporal startToCloseTimeout is set
// to maxRunTimeSeconds + 5 minutes so Temporal only kills the activity if the
// in-process abort mechanism itself fails.
const DEFAULT_MAX_RUN_TIME_SECONDS = 2 * 60 * 60;

const act = proxyActivities<typeof allActivities>({
  startToCloseTimeout: "30 seconds",
  retry: { maximumAttempts: 3 },
});

const actNoRetry = proxyActivities<typeof allActivities>({
  startToCloseTimeout: "60 seconds",
  retry: { maximumAttempts: 1 },
});

const { CreateDraftPR } = proxyActivities<typeof allActivities>({
  startToCloseTimeout: "5 minutes",
  retry: { maximumAttempts: 1 },
});

const { WatchPRState } = proxyActivities<typeof allActivities>({
  startToCloseTimeout: "168 hours",
  heartbeatTimeout: "2 minutes",
  retry: { maximumAttempts: 1 },
});

const { PushFeedbackCommit, RespondToReview } = proxyActivities<typeof allActivities>({
  startToCloseTimeout: "10 minutes",
  retry: { maximumAttempts: 3 },
});

export const wontDoSignal = defineSignal("wont-do");
export const yakStatusQuery = defineQuery<YakWorkflowState>("yak_status");

export async function YakWorkflow(
  yakName: string,
  cfg: WorkflowConfig,
): Promise<string> {
  const state: YakWorkflowState = {
    yakName,
    phase: "init",
    workspace: "",
    prUrl: "",
    prNumber: 0,
    feedbackRound: 0,
  };

  let wontDo = false;
  setHandler(wontDoSignal, () => { wontDo = true; });
  setHandler(yakStatusQuery, () => ({ ...state }));

  // Set initial search attributes so the workflow is findable by yak name.
  upsertSearchAttributes({ YakName: [yakName], Phase: [state.phase] });

  const setPhase = (phase: string) => {
    state.phase = phase;
    upsertSearchAttributes({ Phase: [phase] });
  };

  // Build a RunAgent proxy whose startToCloseTimeout is derived from the
  // configured maxRunTimeSeconds (+ 5 min buffer so Temporal only fires if the
  // in-process abort mechanism itself fails).
  const agentMaxSecs = cfg.pi.maxRunTimeSeconds ?? DEFAULT_MAX_RUN_TIME_SECONDS;
  const agentTimeoutSecs = agentMaxSecs + 5 * 60;
  const { RunAgent } = proxyActivities<{
    RunAgent(yakName: string, workspaceName: string, cfg: PiConfig): Promise<void>;
  }>({
    startToCloseTimeout: `${agentTimeoutSecs} seconds`,
    heartbeatTimeout: "2 minutes",
    retry: { maximumAttempts: 1 },
  });

  // 1. Claim the yak.
  setPhase("claiming");
  await act.YakClaim(yakName);

  try {
    // 2. Init workspace — fresh git clone on a new branch.
    setPhase("init-workspace");
    state.workspace = await actNoRetry.InitWorkspace(cfg.repoUrl, yakName);

    try {
      // 3. Run Pi agent.
      setPhase("implementing");
      await RunAgent(yakName, state.workspace, cfg.pi);

      if (wontDo) {
        return `yak ${yakName} marked won't-do during implementation`;
      }

      // 4. Create draft PR.
      setPhase("creating-pr");
      const pr: PRResult = await CreateDraftPR(cfg.repoUrl, state.workspace, yakName);
      state.prUrl = pr.prUrl;
      state.prNumber = pr.prNumber;
      upsertSearchAttributes({ PrUrl: [pr.prUrl] });

      act.WritePRToYak(yakName, pr.prUrl).catch(() => {});

      // 5. PR feedback loop: wait for merge, close, or REQUEST_CHANGES.
      setPhase("waiting-for-merge");
      let lastReviewId = 0;

      while (true) {
        if (wontDo) {
          await act.YakRelease(yakName, "marked won't-do");
          return `yak ${yakName} marked won't-do`;
        }

        let watchResult: PRWatchResult | undefined;
        WatchPRState(pr.prNumber, cfg.repoUrl, lastReviewId)
          .then((r) => { watchResult = r; })
          .catch(() => { watchResult = { outcome: "closed" }; });
        await condition(() => wontDo || watchResult !== undefined);

        if (wontDo) {
          await act.YakRelease(yakName, "marked won't-do");
          return `yak ${yakName} marked won't-do`;
        }

        const result = watchResult!;

        if (result.outcome === "merged") {
          break;
        }

        if (result.outcome === "closed") {
          await act.YakRelease(yakName, "PR closed without merge");
          return `yak ${yakName} PR closed without merge`;
        }

        // outcome === "feedback"
        const reviewId = result.reviewId!;
        const reviewBody = result.reviewBody ?? "";
        const reviewComments = result.reviewComments ?? [];

        // Build feedback context for the agent.
        const feedbackLines: string[] = [
          `## Review Feedback (round ${state.feedbackRound + 1})`,
          "",
          reviewBody,
        ];
        if (reviewComments.length > 0) {
          feedbackLines.push("", "### Per-line comments");
          for (const c of reviewComments) {
            const loc = c.line !== null ? `:${c.line}` : "";
            feedbackLines.push(`**${c.path}${loc}**: ${c.body}`);
          }
        }
        const feedbackContext = feedbackLines.join("\n");

        setPhase("addressing-feedback");

        // Build a RunAgent proxy for the feedback run.
        const feedbackAgentMaxSecs = cfg.pi.maxRunTimeSeconds ?? DEFAULT_MAX_RUN_TIME_SECONDS;
        const feedbackAgentTimeoutSecs = feedbackAgentMaxSecs + 5 * 60;
        const { RunAgent: RunAgentFeedback } = proxyActivities<{
          RunAgent(yakName: string, workspaceName: string, cfg: PiConfig, feedbackContext?: string): Promise<void>;
        }>({
          startToCloseTimeout: `${feedbackAgentTimeoutSecs} seconds`,
          heartbeatTimeout: "2 minutes",
          retry: { maximumAttempts: 1 },
        });

        await RunAgentFeedback(yakName, state.workspace, cfg.pi, feedbackContext);

        await PushFeedbackCommit(cfg.repoUrl, state.workspace, state.feedbackRound + 1);

        await RespondToReview(
          cfg.repoUrl,
          pr.prNumber,
          reviewId,
          `🪡 Feedback addressed in round ${state.feedbackRound + 1}. Please re-review.`,
        );

        lastReviewId = reviewId;
        state.feedbackRound += 1;
        setPhase("waiting-for-merge");

        // Reset watch result for next iteration.
        watchResult = undefined;
      }

    } finally {
      actNoRetry.CleanupWorkspace(state.workspace).catch(() => {});
    }

  } catch (err) {
    if (state.phase !== "done") {
      await act.YakRelease(yakName, `workflow interrupted at phase ${state.phase}`).catch(() => {});
    }
    throw err;
  }

  // 6. Close the yak.
  setPhase("done");
  await act.YakMarkDone(yakName);

  return `yak ${yakName} done (PR ${state.prUrl})`;
}
